docs/Style.thy

```diff
@@ -0,0 +1,263 @@
+  * definition and fun should be on the same line as the name.
+  * Everything after the first line of a definition or fun statement should be indented.
+  * Type definitions (record, type_synonym, datatype) have the defined type on the same line, the
+    rest only if it fits.
+  * Avoid long hanging indentation in data types, case expressions, etc. That is:
+
+Wrong:
+
+@{term\<open>
+case xs_blah of [] \<Rightarrow> foo
+              | y#ys \<Rightarrow> baz
+\<close>}
+
+Right:
+
+@{term\<open>
+case xs_blah of
+    [] \<Rightarrow> foo
+  | y#ys \<Rightarrow> baz
+\<close>}
+\<close>
+
+text \<open>
+  If a definition body causes an overflow of the 100 char line limit, use one
+  of the following two patterns.
+\<close>
+
+definition hd_opt :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt \<equiv> case_list
+              None
+              (\<lambda>h t. Some h)"
+
+definition hd_opt4 :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt4 \<equiv>
+   case_list None (\<lambda>h t. Some h)"
+
+text \<open>
+  If an apply line causes an overflow of the 100 char line limit, use one
+  of the following two patterns: aligning on ":" or aligning on left.
+\<close>
+
+lemma test_lemma0:
+  "\<lbrakk>hd_opt l \<noteq> None; hd_opt (hd l) = Some x\<rbrakk> \<Longrightarrow>
+    hd_opt (concat l) = Some x"
+  apply (clarsimp simp: hd_opt_def
+                 split: list.splits)
+  done
+
+lemma test_lemma1:
+  "\<lbrakk>hd_opt l \<noteq> None; hd_opt (hd l) = Some x\<rbrakk> \<Longrightarrow>
+    hd_opt (concat l) = Some x"
+  apply (clarsimp simp: hd_opt_def
+                  split: list.splits)
+  done
+
+lemma test_lemma3:
+  "case_option None hd_opt (hd_opt l) = Some x \<Longrightarrow>
+   hd_opt (concat l) = Some x"
+  apply ((cases l; simp add: hd_opt_def),
+         rename_tac h t,
+         case_tac h; simp)
+  done
+
+section \<open>Other\<close>
+
+text \<open>
+General layout:
+* Wrap at 100, avoid overly long lines, not everyone has a big screen
+* No tabs, use spaces instead, standard indentation is 2
+* Avoid starting lines with quotes ("), it breaks some automated tools.
+* Avoid unnecessary parentheses, unless it becomes really unclear what is going on. Roughly,
+  parentheses are unnecessary if Isabelle doesn't show them in output.
+
+Other:
+* Avoid commands that produce "legacy" warnings. Add a JIRA issue with tag cleanup if you see them
+  after an Isabelle update.
+\<close>
+
+section \<open>Comments\<close>
+
+text \<open>
+(* .. *) style comments are not printed in latex. Use them if you don't think anyone will want to
+see this printed. Otherwise, prefer text \<open> .. \<close>
+
+In stark difference to the Isabelle distribution, we do comment stuff. If you have defined or done
+anything that is tricky, required you to think deeply, or similar, save everyone else the duplicated
+work and document why things are the way you've written them.
```

Comment:
Good. I might add something like "That way even if your tactic/definition breaks, we will have an idea of what you were trying to achieve."